### PR TITLE
Deprovision pane for apps

### PR DESCRIPTION
**Pull request by mixonic**
You can deprovision (destroy) an app. @fancyremarker this just destroys the app, let me know if this should be doing anything more fancy.

Success takes you back to the stack view.

### CODE REVIEW

File: app/adapters/application.js [hunk 1]
```diff
@@ -35,6 +35,12 @@ export default HalAdapter.extend({
     } else {
       return error;
     }
+  },
+
+  deleteRecord: function(store, type, record){
```
Comment (mixonic) [inline app/adapters/application.js]: @bantic I think this should be all that is needed to force deletes to hit the non-nested route. I just don't pass the record into `buildURL`, so there is no nesting.

**Comment by fancyremarker** [comment]
Ah — this should actually be creating an operation on the app with `{ type: "deprovision" }` (no other arguments necessary.

**Comment by mixonic** [comment]
@fancyremarker should it also destroy? We can chat over it in detail this afternoon and fix the logic before this merges.

**Comment by fancyremarker** [comment]
Negative. Once the operation is executed by Sweetness, Sweetness will take care of destroying the app.

The issue here is that there are a bunch of resources that need to be cleaned up for an app (app containers, proxy containers, logging containers, Elastic Load Balancer(s), etc.) and so Sweetness needs to do that before the app can actually be deleted. Moreover, we currently have different policies for exactly when the `deprovision` operation gets handled by Sweetness: development apps are deprovisioned immediately, while production apps sit on the queue until manually reviewed.

In the final version of our API, a non-Sweetness user won't even be able to call `DELETE /apps/:id` but that's [a separate issue](https://github.com/aptible/api.aptible.com/issues/103).

File: app/adapters/application.js (hunk 1 above)
Comment (mixonic) [inline app/adapters/application.js]: I don't actually use this deletion code anymore. Instead this flow goes through deprovision

**Comment by mixonic** [comment]
Ready for review by @bantic. Works best with https://github.com/aptible/api.aptible.com/pull/160

File: app/app/deprovision/route.js [hunk 2]
```diff
@@ -0,0 +1,29 @@
...
+      var store = this.store;
+      controller.set('error', null);
+
+      app.get('stack').then(function(stack) {
+        var op = store.createRecord('operation', {
+          type: 'deprovision',
+          app: app
+        });
+        return op.save().then(function(){
+          return app.reload();
```
Comment (bantic) [inline app/app/deprovision/route.js]: this is to pick up the fact that the status of the app is now 'deprovisioning', yeah? Is that info shown in the UI anywhere?
Not strictly necessary, but the intent of the [app model STATUSES](https://github.com/aptible/diesel.aptible.com/blob/master/app/models/app.js#L4-L7) variable is to enumerate all the possible valid statuses. Might be worth adding "deprovisioning" there.

File: app/app/deprovision/template.hbs [hunk 3]
```diff
@@ -1 +1,35 @@
...
+
+          <p>
+            This action CANNOT be undone. This will permanently deprovision
+            <strong>{{model.handle}}</strong>
+          </p>
+          <p>
+            <strong>Type the name of the app to permanently deprovision.</strong>
+          </p>
+          <div class="form-group">
+            {{#button-with-confirmation confirmValue=model.handle action='deprovision'}}
```
Comment (bantic) [inline app/app/deprovision/template.hbs]: this should probably have a class of "btn-danger" to make it red (as shown in sketch: http://take.ms/umRpm)

File: app/components/button-with-confirmation/component.js [hunk 4]
```diff
@@ -0,0 +1,18 @@
+import Ember from 'ember';
+
+export default Ember.Component.extend({
+  isConfirmed: null,
+  isUnconfirmed: Ember.computed.not('isConfirmed'),
+  input: function(){
+    var enteredValue = this.get('enteredValue');
+    var confirmValue = this.get('confirmValue');
+    this.set('isConfirmed', enteredValue === confirmValue);
+  }.on('init'),
```
Comment (bantic) [inline app/components/button-with-confirmation/component.js]: Might be clearer to make `isConfirmed` a cp?

```
isConfirmed: function(){
  return this.get('enteredValue') === this.get('confirmValue');
}.property('enteredValue', 'confirmValue')
```

I had actually forgotten that `input` was an event handler, so I didn't realize what was happening here.

File: tests/acceptance/apps-deprovision-test.js [hunk 5]
```diff
@@ -0,0 +1,108 @@
...
+      type: 'deprovision'
+    });
+  });
+
+  stubStacks();
+
+  signInAndVisit('/apps/'+appId+'/deprovision');
+  fillIn('input[type=text]', appName);
+  andThen(function(){
+    $('input[type=text]').trigger('input');
```
Comment (bantic) [inline tests/acceptance/apps-deprovision-test.js]: If the `button-with-confirmation`'s `isConfirmed` is a CP, this manual trigger appears to not be necessary.

**Comment by bantic** [comment]
awesome — LGTM. A few notes, merge at will.

**Comment by mixonic** [comment]
@bantic This component is definitely a little different than what we've sometimes done in the past. I wrote it experimenting with the idea of data-down actions-up, thus my `input` driven logic instead of a CP dependent on the input value. I'd like to keep experimenting to see if we can make this pattern scale up.

**Pull request merged**